From 584ef3a134a53db62a7d5b0e8ad7672be1322271 Mon Sep 17 00:00:00 2001
From: Stephane Viau <sviau@codeaurora.org>
Date: Tue, 14 Jul 2015 10:28:18 -0400
Subject: [PATCH 30/44] drm/msm/hdmi: Enable HDMI HPD on MTP8994

"qcom,hdmi-tx-hpd"'s function has been changed from apq8084 to msm8996.
apq8084: it's a gpio pin to receive HPD signal.
msm8996: it's the replacement for hpd-5v-en-supply.
So in hdmi_connector code, we can't use this pin as an alternative way to
detect HPD status and it needs to be configured as OUTPUT.

Change-Id: If2aa75c43908297cc086ab226a970ae864ab82bb
Signed-off-by: Jilai Wang <jilaiw@codeaurora.org>
Signed-off-by: Stephane Viau <sviau@codeaurora.org>
---
 drivers/gpu/drm/msm/hdmi/hdmi_connector.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/msm/hdmi/hdmi_connector.c b/drivers/gpu/drm/msm/hdmi/hdmi_connector.c
index 88c727c..1a12eef 100644
--- a/drivers/gpu/drm/msm/hdmi/hdmi_connector.c
+++ b/drivers/gpu/drm/msm/hdmi/hdmi_connector.c
@@ -110,8 +110,7 @@ static int gpio_config(struct hdmi *hdmi, bool on)
 				"HDMI_HPD", config->hpd_gpio, ret);
 			goto error3;
 		}
-		gpio_direction_input(config->hpd_gpio);
-		gpio_set_value_cansleep(config->hpd_gpio, 1);
+		gpio_direction_output(config->hpd_gpio, 1);
 
 		if (config->mux_en_gpio != -1) {
 			ret = gpio_request(config->mux_en_gpio, "HDMI_MUX_EN");
@@ -361,6 +360,13 @@ static enum drm_connector_status hdmi_connector_detect(
 	enum drm_connector_status stat_gpio, stat_reg;
 	int retry = 20;
 
+	/*
+	 * Since the GPIO PIN has been changed from 8084 to 8996, we should
+	 * need to detect the register but not rely on GPIO. This need to be
+	 * refactored later to use DTS configuration.
+	 */
+	return detect_reg(hdmi);
+
 	do {
 		stat_gpio = detect_gpio(hdmi);
 		stat_reg  = detect_reg(hdmi);
@@ -425,8 +431,6 @@ static int hdmi_connector_mode_valid(struct drm_connector *connector,
 				 struct drm_display_mode *mode)
 {
 	struct hdmi_connector *hdmi_connector = to_hdmi_connector(connector);
-	struct hdmi *hdmi = hdmi_connector->hdmi;
-	const struct hdmi_platform_config *config = hdmi->config;
 	struct msm_drm_private *priv = connector->dev->dev_private;
 	struct msm_kms *kms = priv->kms;
 	long actual, requested;
-- 
2.8.2

